src/storage/v2/disk/storage.cpp

```diff
@@ -547,54 +547,126 @@ VerticesIterable DiskStorage::DiskAccessor::Vertices(LabelId label, View view) {
     return VerticesIterable(disk_storage->edge_import_mode_cache_->Vertices(label, view, storage_, &transaction_));
   }
 
-  transaction_.index_storage_.emplace_back(std::make_unique<utils::SkipList<storage::Vertex>>());
-  auto &indexed_vertices = transaction_.index_storage_.back();
-  transaction_.index_deltas_storage_.emplace_back();
-  auto &index_deltas = transaction_.index_deltas_storage_.back();
+  utils::SkipList<memgraph::storage::Vertex> *indexed_vertices{nullptr};
+  std::list<storage::Delta> *index_deltas{nullptr};
 
-  auto gids = MergeVerticesFromMainCacheWithLabelIndexCache(label, view, index_deltas, indexed_vertices.get());
-  LoadVerticesFromDiskLabelIndex(label, gids, index_deltas, indexed_vertices.get());
+  auto merge_with_main_cache = [&](auto &index, auto &index_delta_storage) -> std::unordered_set<Gid> {
+    indexed_vertices = &index[label];
+    index_delta_storage.emplace_back();
+    index_deltas = &index_delta_storage.back();
 
+    return MergeVerticesFromMainCacheWithLabelIndexCache(label, view, *index_deltas, *indexed_vertices);
+  };
+
+  auto &cache = transaction_.label_index_cache_;
+  auto &cache_ci = transaction_.label_index_cache_ci_;
+  auto &index_delta_storage = transaction_.index_deltas_storage_;
+
+  SyncDeletedVertices(cache, label, index_delta_storage, view, merge_with_main_cache);
+
+  if (transaction_.command_id > cache_ci) {
+    cache_ci = transaction_.command_id;
+  } else {
+    if (cache.contains(label)) {
+      if (view == View::OLD) {
+        // TODO
+        // we do not need this merge if we can make sure that
+        // the removing and re-adding the same label to a given
+        // vertex within one transaction is not permitted.
+        merge_with_main_cache(cache, index_delta_storage);
+        return VerticesIterable(AllVerticesIterable(cache[label].access(), storage_, &transaction_, view));
+      }
+      cache[label] = utils::SkipList<Vertex>();
+      merge_with_main_cache(cache, index_delta_storage);
+      return VerticesIterable(AllVerticesIterable(cache[label].access(), storage_, &transaction_, view));
+    }
+
+    cache[label] = utils::SkipList<Vertex>();
+    index_delta_storage.emplace_back();
+    index_deltas = &index_delta_storage.back();
+    LoadVerticesFromDiskLabelIndex({}, label, *index_deltas, transaction_.vertices_);
+  }
+
+  merge_with_main_cache(cache, index_delta_storage);
   return VerticesIterable(AllVerticesIterable(indexed_vertices->access(), storage_, &transaction_, view));
```

Comment:
```suggestion
auto merge_with_main_cache = [&](auto &index, auto &index_delta_storage) -> std::unordered_set<Gid> {
    indexed_vertices = &index[label];
    index_deltas = &index_delta_storage.emplace_back();
    return MergeVerticesFromMainCacheWithLabelIndexCache(label, view, *index_deltas, *indexed_vertices);
  };

  auto &cache = transaction_.label_index_cache_;
  auto &cache_ci = transaction_.label_index_cache_ci_;
  auto &index_delta_storage = transaction_.index_deltas_storage_;

  SyncDeletedVertices(cache, label, index_delta_storage, view, merge_with_main_cache);

  if (transaction_.command_id > cache_ci) {
    cache_ci = transaction_.command_id;
    merge_with_main_cache(cache, index_delta_storage);
    return VerticesIterable(AllVerticesIterable(indexed_vertices->access(), storage_, &transaction_, view));
  }
  if (cache.contains(label)) {
    if (view == View::NEW) {
      cache[label] = utils::SkipList<Vertex>();      
    }
    // TODO
    // we do not need this merge if we can make sure that
    // the removing and re-adding the same label to a given
    // vertex within one transaction is not permitted.
    merge_with_main_cache(cache, index_delta_storage);
    return VerticesIterable(AllVerticesIterable(cache[label].access(), storage_, &transaction_, view));
  }
  cache[label] = utils::SkipList<Vertex>();
  index_deltas = &index_delta_storage.emplace_back();
  LoadVerticesFromDiskLabelIndex({}, label, *index_deltas, transaction_.vertices_);
  merge_with_main_cache(cache, index_delta_storage);
  return VerticesIterable(AllVerticesIterable(indexed_vertices->access(), storage_, &transaction_, view));
```
I am trying to find a way to refactor this to make it a bit better... I think this should work. Just put that comment (TODO) here somewhere

Reply:
I did like this for several reasons:
1. you have explicit steps to see what is going on to analyze specific use case
2. I think it is better to have return in every if case then just one at the end...

Let me know if I misunderstood the logic here